src/java.base/share/classes/sun/security/util/DisabledAlgorithmConstraints.java

```diff
@@ -101,6 +102,8 @@ private static class JarHolder {
 
     private final Set<String> disabledAlgorithms;
     private final Constraints algorithmConstraints;
+    private volatile SoftReference<Map<String, Boolean>> cacheRef =
```

Comment:
It looks like a one-clear-for-all mechanism.  Once the clearing happens, the full map will be collected.  For SoftReferences, it clears them fairly eagerly.  Maybe, the performance could be further improved in practice by using soft reference for each map entry (See sun.security.util.Cache code for an example). 

I will have another look next week.

Reply:
Right, soft references are likely to be cleaned if they are not used in an entire GC cycle.
Using a soft reference for each map entry would not help here; note that all keys and all values in this map are GC roots (keys are enum names, values are `TRUE` and `FALSE`), so in practice they would never be collected.
Even if we made the entries collectable, it would not help with the TLS case; SSLEngine & SSLSocket only check the constraints once at the beginning of the handshake, so they would all expire at the same time anyway. What's even worse, we would then have to clear the expired entries from the map.

I also considered limiting the size of the map. That's a tricky subject; we would need to get the size limit exactly right. Given that the algorithms are always checked in the same order, if the cache were too small, we would get zero hits.
With all the above in mind I decided not to use `sun.security.util.Cache` here.

Reply:
`SoftReference`s are guaranteed to survive one GC after use; beyond that their lifespan is determined by `SoftRefLRUPolicyMSPerMB` and the amount of memory available.

Reply:
> With all the above in mind I decided not to use `sun.security.util.Cache` here

I was not meant to use Cache and timeout for this update.

SoftReference and  this patch should work good in larger memory boxes.  Performance improvement sometimes is a trade-off game between memory and CPU. Did you have a chance to check the performance in the limited memory circumstance?  like '-mx64M".

Reply:
I run a few tests:
`-Xmx16m`, with this patch, still better than the original version:
```
Benchmark                 (resume)  (tlsVersion)   Mode  Cnt     Score     Error  Units
SSLHandshake.doHandshake      true       TLSv1.2  thrpt    5  4477.444 ± 375.975  ops/s
SSLHandshake.doHandshake      true           TLS  thrpt    5   681.471 ±  72.531  ops/s
SSLHandshake.doHandshake     false       TLSv1.2  thrpt    5   335.366 ±  89.056  ops/s
SSLHandshake.doHandshake     false           TLS  thrpt    5   308.711 ±  90.134  ops/s
```

`-Xmx8m`, before patch:
```
Benchmark                 (resume)  (tlsVersion)   Mode  Cnt    Score    Error  Units
SSLHandshake.doHandshake      true       TLSv1.2  thrpt    5  153.760 ± 12.025  ops/s
SSLHandshake.doHandshake      true           TLS  thrpt    5   70.700 ±  7.506  ops/s
SSLHandshake.doHandshake     false       TLSv1.2  thrpt    5   67.459 ±  4.325  ops/s
SSLHandshake.doHandshake     false           TLS  thrpt    5   64.695 ±  1.647  ops/s
```
after:
```
Benchmark                 (resume)  (tlsVersion)   Mode  Cnt    Score    Error  Units
SSLHandshake.doHandshake      true       TLSv1.2  thrpt    5  557.324 ± 50.269  ops/s
SSLHandshake.doHandshake      true           TLS  thrpt    5  155.258 ± 13.866  ops/s
SSLHandshake.doHandshake     false       TLSv1.2  thrpt    5  181.755 ± 29.319  ops/s
SSLHandshake.doHandshake     false           TLS  thrpt    5  120.627 ± 25.832  ops/s
```
Much slower, but still faster with the patch.
With `-Xmx4m` the test failed with OOM.

Reply:
Thanks for the additional testing.  The improvement is impressive.